### ISSUE TITLE
Confusion as to what I am supposed to do in exercise 2.5

### ISSUE THREAD

**Issue by frerich** [mightybyte/monad-challenges#8]
I was quite lost in exercise 2.5, I didn't understand what to do. Given that only a few moments ago I had to define `queryGeek2` in terms of `link`, it seemed easy enough to define `addSalaries` in terms of the earlier defined `lookupMay` and `link` functions:

```
addSalaries :: [(String, Integer)] -> String -> String -> Maybe Integer
addSalaries sals n m =
    lookupMay n sals `link` \salary1 ->
    lookupMay m sals `link` \salary2 ->
    Just (salary1 + salary2)
```

This was the first time I _needed_ the hint to figure out that I was supposed to implement

```
yLink :: (a -> b -> c) -> Maybe a -> Maybe b -> Maybe c
```

However, shouldn't this function maybe be called `yChain`, given the types of `chain` and `link` defined as part of exercise 2.4:

```
chain :: (a -> Maybe b) -> Maybe a -> Maybe b
link :: Maybe a -> (a -> Maybe b) -> Maybe b
```

Finally, further down in the exercise, the text reads

> Notice that in both addSalaries and yLink you have to construct a Maybe out of thin air kind of like we did in Set 1 with mkGen.

By now, I was pretty sure that my `addSalaries` definition was somehow not the intended definition - I didn't feel that I needed to "construct a Maybe out of thin air". Or does a plain `Just` call count?

**Comment by spearsem** [comment]
In the signature for `chain` we start off with a function that accepts an `a` and returns a `Maybe b`: 

`chain ::  (a -> Maybe b) -> Maybe a -> Maybe b`.

As in the above comment, I feel that the exercise text uses the name `yLink` when really it means `yChain`, so the type signature of `yLink` should be analogous to `chain` but for two arguments instead of one.

So why does `yLink` start off with a function that does _not_ return a `Maybe` type?

Why is it

`yLink :: (a -> b -> c) -> Maybe a -> Maybe b -> Maybe c`

instead of

`yLink :: (a -> b -> Maybe c) -> Maybe a -> Maybe b -> Maybe c`?

If we use the latter of these type signatures, then I believe the comment about needing to make a `Maybe` out of thin air no longer applies, and you wouldn't need a call to `Just` nor any use of `mkMaybe` inside the chain/link stuff. 

Instead, usage of that stuff would be pushed back a level, to whatever point it is that you write the safe version of whatever helper function you need. For example, `chain` allows us to use `headMay` instead of `chain` allowing us to use plain `head` but to have to worry about adding a `Just` inside the definition of `chain` (not for pattern matching, but for the result).

The following might help to resolve my confusion.

I believe that if we go with the type signature I am saying from above, then we will be forced to also write a safe helper function, something like `plusMay :: Num a => a -> a -> Maybe a` that _would_ need to use `mkMaybe`. Armed with `plusMay`, we can basically use the same definition for `addSalaries` that @frerich uses above, except we can `yLink` in the final step rather than explicitly handling the `Maybe`-ness of the result, and we can pass the results of `lookupMay` directly as the arguments to `yLink`, instead of getting them one by one and doing the lambda pattern with them.

If this is incorrect, some comments about why would probably go a long way to help me defeat my confusion.

**Comment by mightybyte** [comment]
@frerich As far as the naming goes, don't read too much into the names I chose.  Maybe yChain would have been a better name, but I didn't really want to focus on names.  I even thought about taking it further and naming all the functions something completely irrelevant, like `apple`, `orange`, `banana`, `cherry`, etc.

And you're right that "construct a Maybe out of thin air" simple refers to `Just`.  It seems too simple, but Haskell does a lot by building a lot of fairly simple things on top of each other.  I wish I had a better way to motivate mkMaybe, but this was the best I could think of.  I'm open to improvements here.

If you needed the hint to understand what I was getting at here don't take that as implying that you're not seeing things well enough.  I think it's perfectly reasonable to need this hint since the idea is kind of hard to explain in plain English.

**Comment by mightybyte** [comment]
Also, your addSalary solution using link is definitely not bad.  The point of this exercise was to come to the slightly more convenient yLink formulation.  In later exercises I think it will become more clear why we were doing this.

**Comment by jafristr** [comment]
FWIW, I was unable to come up with the insight - looking at @frerich's solution got me to where I could do the same thing with yLink.
